OpenEXR Loader/Writer

AnastaZIuk opened the pull request:
**Status**:

- the loader is **done**
- the writer is **done**

Thank you @Crisspl for found bug in my array equation

AnastaZIuk committed:
Prepare source for writing OpenEXR dependency

It hasn't involved any macroc in nor project nor cmake yet

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.h:
```diff
@@ -0,0 +1,39 @@
+
+			uint32_t getSupportedFlags() override { return asset::EWF_NONE; }
+
+			uint32_t getForcedFlags() { return 0u; }
```
binary is always forced

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.h:
```diff
@@ -0,0 +1,39 @@
+
+			uint64_t getSupportedAssetTypesBitfield() const override { return asset::IAsset::ET_IMAGE; }
+
+			uint32_t getSupportedFlags() override { return asset::EWF_NONE; }
```
binary

devshgraphicsprogramming commented on src/irr/asset/IAssetManager.cpp:
```diff
@@ -39,6 +39,10 @@
 #include "irr/asset/CImageLoaderTGA.h"
```
you might want to convert all `_IRR_COMPILE_WITH` into CMake options that then go to a configure file

AnastaZIuk replied:
sure, I will deal with it after having openexr loader done

AnastaZIuk committed:
Begin to create OpenEXR base following it's documentation

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -187,6 +187,10 @@ option(IRR_BUILD_DOCS "Enable building documentation?" ON)
 
+unset(OPENEXR_BUILD_UTILS [OPENEXR]) #disable openexr examples
+unset(BUILD_TESTING [BUILD]) #disable openexr bulding tests
+unset(PYILMBASE_ENABLE [PYILMBASE]) #disable PYILMBASE in openexr
```
are you sure this works? To unset a thing I guess it would have to exist first, but you also need to set it before the subfolder is included.

I usually just set the thing to false, right before the subfolder include.

AnastaZIuk replied:
I'm not sure being honest. Well, I will do as you wrote, but I've not known how to do it yet. To fix.

devshgraphicsprogramming replied:
put this in `3rdparty/CMakeLists.txt`

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -29,12 +29,65 @@ namespace irr
+
+			readRgba(fileName, pixels, width, height);
+
+			// TODO put data into buffer, consider OpenEXR 2.0 supported saving layouts 
```
what are `layouts` ?

AnastaZIuk replied:
I meant I need to support tiled files, deep imagesm multilayered images and all stuff I guess you would like it to support:)

devshgraphicsprogramming replied:
it only needs to load regular OpenEXR images produced by mitsuba, and GIMP, you can error out on the more exotic stuff.

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -29,12 +29,65 @@ namespace irr
+				return {};
+
+			readVersionField();
+			readHeader(fileName, ctx);
```
would be cool to re-use this for checking if file can be loaded.

AnastaZIuk replied:
> for checking if file can be loaded.

that will be checked by `magicNumber`, sure

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.h:
```diff
@@ -6,6 +6,20 @@
+namespace IMATH = IMATH_NAMESPACE;
+
+using namespace IMF;
+using namespace IMATH;
```
NEVER PUT `using namespace` in a header!

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.h:
```diff
@@ -32,6 +46,64 @@ namespace irr
+
+			struct SContext
+			{
+				constexpr static std::array<char, 4> magicNumber = { 0x76, 0x2f, 0x31, 0x01 };
```
watch out, magic numbers are usually stored as `uint32_t` for the purpose of endianness check, see my SPIRV loader

AnastaZIuk committed:
Change OpenEXR submodule to 2.4 tag

Crisspl committed:
Merge pull request #20 from buildaworldnet/shader_pipeline

Fix the IDescriptorSet clusterfuck.

AnastaZIuk committed:
Add checking magicNumber, correct Cmake - still need to be fixed!

AnastaZIuk commented on source/Irrlicht/CMakeLists.txt:
```diff
@@ -212,21 +216,34 @@ macro(irr_target_link_shaderc _trgt)
+	target_include_directories(${_trgt} PUBLIC ${THIRD_PARTY_SOURCE_DIR}/libpng)
+endmacro()
+macro(irr_target_link_openexr _trgt)
+	#add_dependencies(${_trgt} openexr)
```
I wanted to add dependency as IlmBase or OpenEXR, that was my mistake, there is no need to add it since the only one dependency that could be added is `zlib`, but it is already done earlier

devshgraphicsprogramming replied:
they're adding their own zlib I guess.

you need to add dependency IlmBase, etc. just as jpeg, png, shaderc are adde above
then link libraries,
then add include directories

seems you'll need to add dependency and link IlmBase as well

devshgraphicsprogramming replied:
I've looked through your log, you're getting compile errors before you're getting linking errors, hence you don't know if a library called `openexr` even exists (wont get messages about library not found)

Also your log jumps straight from Documetnation target to IRrlicht, there are no Ilm or OpenEXR targets being built, this tells me that IRrlicht does not have the Ilm targets in the dependency chain.

*Before*
Introduce `_IRR_COMPILE_WITH_OPENEXR_LOADER_` and disable it.
See if you can compile with the loader disabled, if you can't you've fucked up an exisitng and working build system.

*First*
try to add dependencies explicitly, so you see the Ilm/OpenEXR targets being built or checked for building in the log

*Second*
Fix your missing include issues (include dirs), but make sure to sort out dependencies above first, as the build process may be producing include files in the build dir that you need to find.

*Third*
Maybe limit the OpenEXR includes to the .cpp file, so that the whole project doesnt get polluted by OpenEXR declarations.

AnastaZIuk replied:
_status:_

- Before - **DONE properly, projects build**
- First - ...
- Second - ...
- Third - ...

devshgraphicsprogramming replied:
what the progress on the first?

devshgraphicsprogramming replied:
I did it all for you , see PR to your repo

devshgraphicsprogramming commented on 3rdparty/CMakeLists.txt:
```diff
@@ -79,6 +80,9 @@ option(PNG_TESTS  "Build libpng tests" OFF)
 
+#openexr
+add_subdirectory(openexr openexr EXCLUDE_FROM_ALL)
+
```
here you must configure your lib, add all the options, build static, etc. just before

AnastaZIuk replied:
shouldn't I configure lib in source CMakeLists?

devshgraphicsprogramming replied:
no, you must set up all the CMake options for the subfolder before you include it.

AnastaZIuk committed:
Deal with cmake, first attempt

Crisspl committed:
jpeg loader

devshgraphicsprogramming committed:
Merge branch 'openexr' into OpenEXR

devshgraphicsprogramming committed:
Merge pull request #407 from AnastaZIuk/OpenEXR

Let me try the cmake

devshgraphicsprogramming committed:
fixed CMake for OpenEXR

devshgraphicsprogramming committed:
Remove .X loader (will do glTF in the future)

devshgraphicsprogramming committed:
remove X Loader

devshgraphicsprogramming committed:
finish removing .x

devshgraphicsprogramming committed:
add more useful common asset-metadata API

AnastaZIuk committed:
Merge pull request #8 from buildaworldnet/openexr

Openexr

AnastaZIuk committed:
Add Version field support

AnastaZIuk committed:
Add buffer handling

AnastaZIuk committed:
Fix an error, link openexr to IrrlichtServer

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -41,17 +41,69 @@ namespace irr
-			// TODO put data into buffer, consider OpenEXR 2.0 supported saving layouts 
-			return SAssetBundle{};
+			constexpr uint32_t MAX_PITCH_ALIGNMENT = 8u;									   // OpenGL cannot transfer rows with arbitrary padding
+			auto calcPitchInBlocks = [](uint32_t width, uint32_t blockByteSize) -> uint32_t	   // try with largest alignment first
```
this function is being used in more than one image loader, maybe we should make it into a static method of something common?

AnastaZIuk replied:
OpenEXR uses it as well, it's a good idea I think

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -41,17 +41,69 @@ namespace irr
+			region.imageSubresource.baseArrayLayer = 0u;
+			region.imageSubresource.layerCount = 1u;
+			region.bufferOffset = 0u;
+			region.bufferRowLength = calcPitchInBlocks(width, getTexelOrBlockBytesize(image->getCreationParameters().format));
```
you need to take alignment into account as well, given by calcPitchInBlocks ()

Think of it as a fake/padded width.

AnastaZIuk replied:
isn't it good? (outdated comment?)

devshgraphicsprogramming replied:
calcPitchInBlocks doesn't take alignment into account

AnastaZIuk replied:
well, I checked it and this is a copy of PNG loader `calcPitchInBlocks`
what did I miss?

AnastaZIuk replied:
well, you mean alpha channel right?

devshgraphicsprogramming replied:
pitchInBlocks is a fake "length", as in while the image can be a 200x200 you can make the length 252.

This only affects the buffer unpack to image when making a GL texture from a GL Buffer.

This means that the following equation to compute the output address is wrong
https://github.com/buildaworldnet/IrrlichtBAW/pull/403/files#diff-63315eefc23fcfd4f83552a09eee0c58R240

This is why PNG Loader uses `getPitch` or something like that instead of the image width, beause of the possible padding at the end of each row when storing image data in a buffer.

AnastaZIuk replied:
to fix

AnastaZIuk replied:
so assume I'm handling 250th pixel, what should I fill it with if it's fake actually?
> pitchInBlocks is a fake "length", as in while the image can be a 200x200 you can make the length 252.

AnastaZIuk replied:
I'm filling those fake values with 0

devshgraphicsprogramming replied:
dont fill fake with anything, it wont ever be used.

You just need to change your array output addressing code, from `x+y*realWidth` to `x+y*bufferRowLength`

AnastaZIuk replied:
great

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -41,17 +41,69 @@ namespace irr
+			region.imageOffset = { 0u, 0u, 0u };
+			region.imageExtent = image->getCreationParameters().extent;
+
+			float* dataToSend = _IRR_NEW_ARRAY(float, image->getImageDataSizeInBytes());
```
you could do `dataToSend = texelBuffer->getPointer()` instead of wasting a copy.

AnastaZIuk replied:
but it won't be a reference pointer, but simply pointer. Won't it crash?

devshgraphicsprogramming replied:
as long as the ICPUBuffer lives (the smart pointer doesn't go out of scope), the pointer returned by `getPointer` is valid

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -41,17 +41,69 @@ namespace irr
+				dataToSend[i] = pixels[i]->r;
+				dataToSend[i + 1] = pixels[i]->g;
+				dataToSend[i + 2] = pixels[i]->b;
+				dataToSend[i + 3] = pixels[i]->a;
```
doesn't OpenEXR have a function other than `readRgba` that will give you quantized `uint16_t/halfFloat` directly instead of dequantizing into full floats?

Also your image format you declared as RGBA half float above, your buffers are also same size as a half float image, and here you're reading full floats and then memcpying them to buffer. this will be a disaster.

devshgraphicsprogramming replied:
You need to know when file is `ImfRgba` (half) and you need `EF_R16G16B16A16_SFLOAT`
https://github.com/AcademySoftwareFoundation/openexr/blob/1b88251b8d955124d7a5da9716ec287ef78440e5/OpenEXR/IlmImf/ImfCRgbaFile.h#L73

And when we have `Rgba` (full float)  and you need `EF_R32G32B32A32_SFLOAT`
https://github.com/AcademySoftwareFoundation/openexr/blob/625b95fa026c3b78e537e9bb6a39fcd51920ad13/OpenEXR/IlmImf/ImfRgba.h#L55

I don't know if OpenEXR allows for mixed precision channels (R Half, G full) if you find out that it does, then convert to highest one present (if one channel is 32bit, then all 32bit).

AnastaZIuk replied:
As I can see they call RGBA something that uses their `Half`s being 16 bit precision floats

devshgraphicsprogramming replied:
see the struct declarations/definitions, there's one for full floats and one for half.

AnastaZIuk replied:
https://github.com/AcademySoftwareFoundation/openexr/blob/1b88251b8d955124d7a5da9716ec287ef78440e5/OpenEXR/IlmImf/ImfCRgbaFile.h#L50 `IlmHalf`
and `Half` https://github.com/AcademySoftwareFoundation/openexr/blob/625b95fa026c3b78e537e9bb6a39fcd51920ad13/IlmBase/Half/half.h#L41 right? Both seem to be 16 bit precision, but `IlmHalf` is unsigned shrot and `Half` is a 16 bit float (maybe I missed something)

devshgraphicsprogramming replied:
they're all stored as uint16_t, its just that the bits have a different meaning than for an actual `uint16_t`

AnastaZIuk replied:
well, because of that they provide this though, it will be good for me https://github.com/AcademySoftwareFoundation/openexr/blob/1b88251b8d955124d7a5da9716ec287ef78440e5/OpenEXR/IlmImf/ImfConvert.h#L100

devshgraphicsprogramming replied:
you dont' want to convert

an array of IlmHalf is good to do a straight copy to a buffer for an EF_R16G16B16A16_SFLOAT texture

AnastaZIuk committed:
Put OpenEXR headers into .cpp, deal with zlib

AnastaZIuk committed:
Add missing zlib library to Irrlicht project

Crisspl committed:
A little rework on CShaderIntrospector

Crisspl committed:
CShaderIntrospector::createApproximateFromIntrospection() for desc set layout

devshgraphicsprogramming committed:
Merge pull request #409 from AnastaZIuk/OpenEXR

More CMake, yay!

Crisspl committed:
CShaderIntrospector::createApproximateFromIntrospection for push constant rngs

partially done

devshgraphicsprogramming committed:
long and painful fix to the zlib issue (I hate you OpenEXR for using `find_package`)

devshgraphicsprogramming committed:
finally a fix for the pain imposed upon us by CMake (I hate find_package)

AnastaZIuk committed:
Merge pull request #9 from buildaworldnet/openexr

ZLIB CMake final solution

AnastaZIuk committed:
Add handling available OpenEXR formats, fix some issues

AnastaZIuk committed:
Make it all build and compile properly

Crisspl committed:
CShaderIntrospector::createApproximatePipelineLayoutFromIntrospection

Crisspl committed:
CShaderIntrospector: createApproximate* for compute and graphics pipelines

might want to merge them into one function which deduces what pipeline to create on its own, but IMO it's not necessary

AnastaZIuk committed:
Support RGBA and RGB as well

Crisspl committed:
CShaderIntrospector: merged PC ranges + minor fixes

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -0,0 +1,454 @@
+			if (isTheBitActive(10))
+				versionField.doesFileContainLongNames = true;
+			else
+				versionField.doesFileContainLongNames = false;
```
if you dont support something here that would prevent loading a file, feel free to return false (like multi-part file)

AnastaZIuk replied:
sure, to fix!

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -0,0 +1,454 @@
+
+		void specifyIrrlichtEndFormat(E_FORMAT& format, const InputFile& file, bool& doesItSupportAlphaChannel)
+		{
+			const IMF::Channel* RChannel = file.header().channels().findChannel("R");
+			const IMF::Channel* GChannel = file.header().channels().findChannel("G");
+			const IMF::Channel* BChannel = file.header().channels().findChannel("B");
```
there's also XYZ formats (although thats in CIE XYZ colorspace)

AnastaZIuk replied:
there are X Y Z channels, but do we want to support it?

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -0,0 +1,454 @@
+					format = EF_R32G32B32A32_UINT;
+				else
+					format = EF_R32G32B32_UINT;
+			}
```
the alpha checks will cause a texture with a present but different alpha format to RGB to register as a non-alpha texture (should fail instead)

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -0,0 +1,454 @@
+				pixelType = PixelType::UINT;
+
+			for (uint8_t rgbaChannelIndex = 0; rgbaChannelIndex < (doesItSupportAlphaChannel ? 4 : 3); ++rgbaChannelIndex)
+				frameBuffer.insert
```
why are you extracting RGBA into 4 separate planes?

AnastaZIuk replied:
because they provide RGBA class, but don't provide RGB. 
Actually I find it better approach for data assignment and more clear,
because we base on available channels

AnastaZIuk replied:
well, and in that way you can specify Slice to make sure Ilm reads type and handle it correctly

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -0,0 +1,454 @@
+					reinterpret_cast<const char*>(rgbaSignatureAsText[rgbaChannelIndex]),                       // name
+					Slice(pixelType,                                                                            // type
+					(char*)(&(pixelRgbaMapArray[rgbaChannelIndex])[0][0] - dw.min.x - dw.min.y * width),        // base
+					sizeof((pixelRgbaMapArray[rgbaChannelIndex])[0][0]) * 1,                                    // xStride
+					sizeof((pixelRgbaMapArray[rgbaChannelIndex])[0][0]) * width,                                // yStride
```
are you sure the strides are in pixels and not bytes?

AnastaZIuk replied:
everything is done well, I've read a few pages of their documentation

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -0,0 +1,454 @@
+				for (uint64_t yPos = 0; yPos < height; ++yPos)
+					for (uint8_t channelIndex = 0; channelIndex < (doesItSupportAlphaChannel ? 4 : 3); ++channelIndex)
+					{ 
+						const uint64_t ptrStyleEndShiftToImageDataPixel = (yPos * width) + xPos;
```
This is what I refer to in https://github.com/buildaworldnet/IrrlichtBAW/pull/403/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=No+extension&hide-deleted-files=true#r358141441

AnastaZIuk replied:
it was wronged, fixed without aligement

Crisspl committed:
CShaderIntrospector::getImageInfoFromIntrospection()

AnastaZIuk committed:
Fix CMake, fill IrrBuffer properly without "fake alignment"

Crisspl committed:
CShaderIntrospector: extensions as ptr range

AnastaZIuk committed:
Fix RGBA and RGB handling, deal with errors, check buffers

AnastaZIuk committed:
Fix checking magic number

Crisspl committed:
IAsset::clone() impl for all assets

Crisspl committed:
oops

AnastaZIuk committed:
Add OpenEXR writer

AnastaZIuk committed:
Perform some tests

Crisspl committed:
Merge branch 'OpenEXR' of github.com:AnastaZIuk/IrrlichtBAW into shaders

Crisspl committed:
Fixed some errors

AnastaZIuk committed:
Merge pull request #10 from Crisspl/shaders

Shaders

AnastaZIuk committed:
Fix writing, but still weird values

AnastaZIuk committed:
Merge branch 'OpenEXR' of https://github.com/AnastaZIuk/IrrlichtBAW into OpenEXR

AnastaZIuk committed:
Fix everything, deallocation on writing left

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -22,14 +22,91 @@ SOFTWARE.
+			#else
+				dynamic_cast<const asset::ICPUImage*>(_params.rootAsset);
+			#endif
+			assert(image);
```
we should proabably make a macro out of this or a template utility function.

AnastaZIuk replied:
well, so where should it be put in files?

devshgraphicsprogramming replied:
probably a template static function in `IAsset`

devshgraphicsprogramming replied:
call it `castDown`

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -22,14 +22,91 @@ SOFTWARE.
+			else if (params.type == EF_R32G32B32A32_SFLOAT)
+				pixelType = PixelType::FLOAT;
+			else if (params.type == EF_R32G32B32A32_UINT)
+				pixelType = PixelType::UINT;
```
you need an error out else, people might call it with stupid formats images (or Low Dynamic Range images).

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -22,14 +22,91 @@ SOFTWARE.
+			for (const auto& channelAsText : rgbaSignatureAsText)
+				header.channels().insert(channelAsText, Channel(pixelType));
+
+			auto byteSizeOfSingleChannelPixel = params.format == EF_R16G16B16A16_SFLOAT ? 2 : 4;
```
instead of using unintelligible costants such as 2,4 use `sizeof(uint16_t)`,`sizeof(float)`

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -22,14 +22,91 @@ SOFTWARE.
+					rgbaSignatureAsText[rgbaChannelIndex],                                                       // name
+					Slice(pixelType,                                                                             // type
+					(char*) image->getBuffer()->getPointer(),                                                    // base
+					byteSizeOfSingleChannelPixel * 1,                                                            // xStride
```
nope, bad stride, you need stride all pixels at once (so sizeof whole pixel)

devshgraphicsprogramming replied:
you could do the bufferRowLength*pixelsize

AnastaZIuk replied:
Because I insert a layout for one channel. It's the OpenEXR library job to perform appropriate calculations and save it properly. Just save a file, you will see it is correct

AnastaZIuk replied:
and it's outdated

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -22,14 +22,91 @@ SOFTWARE.
+				(
+					rgbaSignatureAsText[rgbaChannelIndex],                                                       // name
+					Slice(pixelType,                                                                             // type
+					(char*) image->getBuffer()->getPointer(),                                                    // base
```
you need a per component offset here, 0 for red, one unit for green, two units for blue, etc.

AnastaZIuk replied:
outdated

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -0,0 +1,159 @@
+			file.writePixels(image->getCreationParameters().extent.height);
+
+			//for (uint8_t channel = 0; channel < availableChannels; ++channel)
+				// _IRR_DELETE_ARRAY(pixelsArrayIlm[channel], width * height); doesn't work (?)
```
what do you mean by doesn't work? doesn't compile?

The macro might not like the pointer type going in (is it a void* ?)

AnastaZIuk replied:
it doesn't compile, and type is known by the template

AnastaZIuk replied:
and it isn't void*

devshgraphicsprogramming replied:
errors from compiler?

AnastaZIuk replied:
![image](https://user-images.githubusercontent.com/34793522/71409137-c69a8800-2640-11ea-9d0b-d7204743ba6b.png)

devshgraphicsprogramming replied:
try making the pointer into a constant pointer and let me know (compiler might get confused).

devshgraphicsprogramming replied:
ok I might have fucked this one up in the following lines
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/core/memory/new_delete.h#L104

I use `std::remove_pointer` but I should also then wrap it in `std::remove_reference` first.

The compiler is moaning about the fact that the type being passed into the templates is a reference to a type (after you put a pointer type into a `std::remove_pointer` you get a reference type, not the straight vanilla type).

So feel free to correct my macros ;)

AnastaZIuk replied:
Macros fixed, but it still doesn't work, but it works with & and works if you create `ilmType* poitner` and try to delete it

AnastaZIuk replied:
for each loop works as well

AnastaZIuk replied:
curious is that if you perform for each loop when an element is a reference pointer while deallocating memory, it spreads out the same errors

devshgraphicsprogramming replied:
so its done or not?

AnastaZIuk replied:
It's done, above shows you the behaviour

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -0,0 +1,159 @@
+
+			OutputFile file(fileName, header);
+			file.setFrameBuffer(frameBuffer);
+			file.writePixels(image->getCreationParameters().extent.height);
```
you're only writing 720 pixels of a 1280x720 image here, just saying.

AnastaZIuk replied:
It's good, the parameter of this specifies how many scan lines it should write
if I were to put there 1, it would save 1280 x 1 image

because above I describe the end layout, it saves it correctly

devshgraphicsprogramming replied:
ah ok. cool

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -0,0 +1,159 @@
+					sizeof(*pixelsArrayIlm[channel]) * 1,                                                        // xStride
+					sizeof(*pixelsArrayIlm[channel]) * image->getCreationParameters().extent.width)              // yStride
+				);
+			}
```
same bugs as before.

I'd actually make a Width X Height "framebuffer" (just an array) and copy regions to it manually as necessary.

AnastaZIuk replied:
it's done well

>  a Width X Height "framebuffer"

that's what it is
pixelsArrayIlm keeps colors matched to single channel
https://github.com/buildaworldnet/IrrlichtBAW/pull/403#discussion_r361109229

AnastaZIuk replied:
important is that I have to insert each channel separately because RGBA class forces you to use half format
that's why I pass to it such an array , to overpass it and insert channels that may have different types (half, full float and uint32_t)

> a Width X Height "framebuffer"

so it's correct

devshgraphicsprogramming replied:
if you set up the offset (pointer), stride (penultimate parameter) and row length (last parameter) properly, you'll be able to basically `memcpy` entire rows.

AnastaZIuk replied:
so actualy I only have to copy to pixelsArrayIlm pixels correctly having regions taken into account

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -0,0 +1,159 @@
+
+			const auto& width = image->getCreationParameters().extent.width;
+			const auto& height = image->getCreationParameters().extent.height;
+			const auto pitch = image->getRegions().begin()->bufferRowLength;
```
this will be potentially different for every region, this is why you need to reserve the correct image sized buffer and copy pixels manually (construct the image from regions)

AnastaZIuk replied:
there is only one region :D

devshgraphicsprogramming replied:
no goddamn it, I can give you an image with 50 regions in the spec and you need to be able to save it!

AnastaZIuk replied:
to fix

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -0,0 +1,159 @@
+						ilmType channelPixel = *(reinterpret_cast<const ilmType*>(image->getBuffer()->getPointer()) + ptrStyleEndShiftToImageDataPixel + channelIndex);
+						*(pixelsArrayIlm[channelIndex] + ptrStyleIlmShiftToDataChannelPixel) = channelPixel;
+					}
+				}
```
wrong, you need to iterate through regions, and then `x,y` need to go from `offset` to `offset+extent` then obviously use all the other region struct parameters to get the pixels out of the buffer.

AnastaZIuk replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/403#discussion_r361111290

AnastaZIuk replied:
and it actually saves a file properly, I've tested it

devshgraphicsprogramming replied:
yeah because the region encompasses the whole image, and you haven't been handed a difficult case yet.

devshgraphicsprogramming commented on include/irr/asset/ICPUImage.h:
```diff
@@ -77,7 +77,7 @@ class ICPUImage final : public IImage, public IAsset
 				return false;
 		
-			buffer = _buffer;
+			buffer = std::move(_buffer);
```
thx for this

devshgraphicsprogramming commented on include/irr/asset/ICPUImage.h:
```diff
@@ -30,10 +30,10 @@ class ICPUImage final : public IImage, public IAsset
         {
             auto par = params;
-            auto cp = core::make_smart_refctd_ptr<ICPUImage>(std::move(par));
+            auto cp = core::smart_refctd_ptr<ICPUImage>(new ICPUImage(std::move(par)), core::dont_grab);
```
why @Crisspl ?

Crisspl replied:
ctor is inaccessible for make_smart...() func

devshgraphicsprogramming replied:
ah ok, then use `create`

AnastaZIuk committed:
Correct some macros, add castDown template function, handle unexpected format in writer

AnastaZIuk committed:
Devide regions into sub-array pixels to copy it to main array pixels

AnastaZIuk committed:
Adjust pixel pulling equation to regions

AnastaZIuk committed:
Add IImageLoader and flattenRegions template function

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -77,44 +77,30 @@ namespace irr
-					width += region->bufferRowLength;
-					height += region->bufferImageHeight;
+					width += region->imageExtent.width;
+					height += region->imageExtent.height;
```
nope, you can get ths from image creation parameters

AnastaZIuk replied:
but won't it affect then with another mipmap levels? 
unless image creation parameters hold extent for the buffer assuming 0 mipmap level as base level for each region

devshgraphicsprogramming replied:
image creation parameters must create an image big enough so all regions in the list are valid (i.e. can fit inside)

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -77,44 +77,30 @@ namespace irr
 						for (uint8_t channelIndex = 0; channelIndex < availableChannels; ++channelIndex)
 						{
-							ilmType channelPixel = *(reinterpret_cast<const ilmType*>(image->getBuffer()->getPointer()) + ptrStyleEndShiftToImageDataPixel + channelIndex);
+							ilmType channelPixel = *(reinterpret_cast<const ilmType*>(data) + ptrStyleEndShiftToImageDataPixel + channelIndex);
```
region also specifies a byte offset into the beginning of the buffer.

AnastaZIuk replied:
I know, but isn't it good? If not, how should I combine a byte offset into the beginning of the buffer with `ptrStyleEndShiftToImageDataPixel`?

AnastaZIuk replied:
Do I need it if I specify a begginging of each pixel using `ptrStyleEndShiftToImageDataPixel`?

AnastaZIuk committed:
Fix handling regions

devshgraphicsprogramming commented on include/irr/asset/IAssetWriter.h:
```diff
@@ -62,7 +59,7 @@ enum E_WRITER_FLAGS : uint32_t
 	@see E_WRITER_FLAGS
 */
-class IAssetWriter : public virtual core::IReferenceCounted
+class IAssetWriter : public virtual IImageLoader
```
WAIT WHAT!?

devshgraphicsprogramming commented on include/irr/asset/IAssetWriter.h:
```diff
@@ -3,10 +3,7 @@
-
-#include "IWriteFile.h"
-#include "IAsset.h"
+#include "irr/asset/IImageLoader.h"
```
THE F***?

AnastaZIuk committed:
Fix fucked inheritance

AnastaZIuk committed:
Fix the pixel pulling equation

devshgraphicsprogramming commented on include/irr/asset/ICPUShader.h:
```diff
@@ -43,7 +56,7 @@ class ICPUShader : public IAsset, public IShader<ICPUBuffer>
 		//! Might be GLSL null-terminated string or SPIR-V bytecode (denoted by m_containsGLSL)
 		core::smart_refctd_ptr<ICPUBuffer>	m_code;
-		const bool							m_containsGLSL;
+		/*const */bool							m_containsGLSL;
```
@Crisspl could still keep it const if you made a new constructor for ICPUShader

devshgraphicsprogramming commented on include/irr/config/BuildConfigOptions.h.in:
```diff
@@ -15,6 +15,12 @@
 #cmakedefine _IRR_COMPILE_WITH_OPENGL_
 #cmakedefine _IRR_COMPILE_WITH_VULKAN_
 
+// loaders
+#cmakedefine _IRR_COMPILE_WITH_OPENEXR_LOADER_
+
+// writers
+#cmakedefine _IRR_COMPILE_WITH_OPENEXR_WRITER_
+
```
@AnastaZIuk can you do this for all loaders (and obviously pick good defaults)

AnastaZIuk replied:
sure!

AnastaZIuk replied:
after PR merging of course, before I handle GLI

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderJPG.cpp:
```diff
@@ -274,20 +291,48 @@ asset::SAssetBundle CImageLoaderJPG::loadAsset(io::IReadFile* _file, const asset
-	uint32_t& width = imageSize[0];
-	uint32_t& height = imageSize[1];
+
+    // OpenGL cannot transfer rows with arbitrary padding
+    static const uint32_t MAX_PITCH_ALIGNMENT = 8u;
+    // try with largest alignment first
+    auto calcPitchInBlocks = [](uint32_t width, uint32_t blockByteSize) -> uint32_t
+    {
+        auto rowByteSize = width * blockByteSize;
+        for (uint32_t _alignment = MAX_PITCH_ALIGNMENT; _alignment > 1u; _alignment >>= 1u)
+        {
+            auto paddedSize = core::alignUp(rowByteSize, _alignment);
+            if (paddedSize % blockByteSize)
+                continue;
+            return paddedSize / blockByteSize;
+        }
+        return width;
+    };
```
why is this not factored out to a common static function for some base class like `IImageLoader` and `IImageWriter`

AnastaZIuk replied:
should I put it into `IImageLoader`? I can handle it as well after merging

devshgraphicsprogramming commented on src/irr/asset/IImageLoader.h:
```diff
@@ -0,0 +1,45 @@
+			public:
+
+				IImageLoader();
+				virtual ~IImageLoader();
```
do a pure virtual destructor, do the constructor as a default

AnastaZIuk replied:
pure virtual destructor, why?

devshgraphicsprogramming replied:
pure virtual destructor is for a class that we never ever intend to instantiate (Abstract Data Type) and that is a pure interface (no members, etc.)

AnastaZIuk replied:
but I need to provide a definition of the destructor, where should I put it?

AnastaZIuk replied:
I need to create for it cpp file however

devshgraphicsprogramming replied:
`virtual ~IImageLoader() =0;` thats a pure virtual destructor, makes the class abstract, the destructor never gets instantiated, so you never need the definition

devshgraphicsprogramming replied:
`IReferenceCounted` has one

devshgraphicsprogramming commented on src/irr/asset/IImageLoader.cpp:
```diff
@@ -0,0 +1,33 @@
+		IImageLoader::~IImageLoader() {}
+
+	}
+}
```
No need for a CPP file for this

AnastaZIuk replied:
it'll be deleted

devshgraphicsprogramming commented on include/irr/asset/IAssetWriter.h:
```diff
@@ -59,7 +61,7 @@ enum E_WRITER_FLAGS : uint32_t
 	@see E_WRITER_FLAGS
 */
-class IAssetWriter : public virtual IImageLoader
+class IAssetWriter : public virtual core::IReferenceCounted, public virtual IImageLoader
```
its an asset writer!!!!

make a `IImageWriter` that inherits from `IAssetWriter`

please return `IAssetWriter` to virtual inheritance from `IReferenceCounted`

devshgraphicsprogramming commented on src/irr/asset/IImageLoader.h:
```diff
@@ -7,15 +7,12 @@
 	{
 
-		class IImageLoader : public virtual core::IReferenceCounted
+		class IImageLoader 
```
it was correct before.

devshgraphicsprogramming commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -86,7 +86,7 @@ namespace irr
 					{
-						const uint64_t ptrStyleEndShiftToImageDataPixel = (yPos * width * availableChannels) + (xPos * availableChannels) + region->bufferOffset;
+						const uint64_t ptrStyleEndShiftToImageDataPixel = (yPos * region->imageExtent.width * availableChannels) + (xPos * availableChannels) + region->bufferOffset;
 						const uint64_t ptrStyleIlmShiftToDataChannelPixel = (yPos * width) + xPos;
```
wrong, general equation is `((z*Height+y)*Width+x)*PixelOrBlockSize+ByteOffset`

Height and Width are not the image extent visual/gometrucal parameters, but buffer jumps:
see https://github.com/buildaworldnet/IrrlichtBAW/blob/shader_pipeline/include/irr/asset/IImage.h#L452

AnastaZIuk replied:
Where did you get the general equation from? Could you give me a reference?

AnastaZIuk replied:
Now it doesn't work

Crisspl replied:
First of all: image texel format size varies, so you cannot iterate over `reinterpret_cast<const ilmType*>(data)`. Iterate bytewise instead. Now your `ptrStyleEndShiftToImageDataPixel` includes `bufferOffset` and then used to increment ptr of type `ilmType*`. So ptr is getting moved by `sizeof(ilmType)*bufferOffset` bytes.
IDK why actually you are iterating over ilmType array since (as i said) channel size varies (half=2bytes or float=4bytes) and ilmType has constant size (cpp is statically typed), but anyway: to get ptr to texel you need to use equation matt gave you. If EXR is always 2D image (i dont know), you can simplify it to `(y*Width+x)*PixelOrBlockSize+ByteOffset`. It's theoretically the same what you have now, but yours is weird because this iterating over `ilmType*` thing. So in practice:
```
void* texelPtr = reinterpret_cast<uint8_t*>(data) + bufferOffset + (yPos * region->imageExtent.width + xPos) * PixelOrBlockSize;
```

Crisspl replied:
ah, ilmType is template param, ok. Anyway everything above applies

AnastaZIuk replied:
Thank you @Crisspl! I have reorganized the code and now it works
I should have perform calculations without using bytewise shifts

devshgraphicsprogramming commented on src/irr/asset/IImageLoader.h:
```diff
@@ -27,7 +27,7 @@ namespace irr
 							for (uint64_t xPos = region->imageOffset.x; xPos < region->imageOffset.x + region->imageExtent.width; ++xPos)
 							{
-								const uint64_t offsetToPixelBeggining = (yPos * rowStride * pixelStride) + (xPos * pixelStride) + region->bufferOffset;
+								const uint64_t offsetToPixelBeggining = (yPos * region->imageExtent.width * pixelStride) + (xPos * pixelStride) + region->bufferOffset;
```
similar here w.r.t 
https://github.com/buildaworldnet/IrrlichtBAW/blob/shader_pipeline/include/irr/asset/IImage.h#L452

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -0,0 +1,464 @@
+
+			if (XChannel && YChannel && ZChannel)
+			{
+				os::Printer::log("LOAD EXR: the file consist of not supported CEI XYZ channels", file.fileName(), ELL_ERROR);
```
CIE no CEI

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderOpenEXR.cpp:
```diff
@@ -0,0 +1,464 @@
+				os::Printer::log("LOAD EXR: the file consist of not supported CEI XYZ channels", file.fileName(), ELL_ERROR);
+				return false;
+			}
+			else if (!RChannel || !GChannel || !BChannel)
```
we can support single R channel, RG channel combo and RGB channel combo

devshgraphicsprogramming replied:
but obvs no GB, G, B, RB etc.

AnastaZIuk replied:
if no channel, what value should I fill it with?

devshgraphicsprogramming replied:
there are R16 and R32 formats as well as RG16 and RG32, RGB is special cause it would be Not Power of Two (NPoT) so it doesnt exists and you use RGBA instead

AnastaZIuk committed:
Adjust to requirements

AnastaZIuk committed:
Fix the equation

AnastaZIuk committed:
Make descructors pure virtual

Crisspl commented on src/irr/asset/CImageWriterOpenEXR.cpp:
```diff
@@ -90,12 +90,12 @@ namespace irr
 					for (uint64_t xPos = region->imageOffset.x; xPos < region->imageOffset.x + regionWidth; ++xPos)
 					{
-						const uint64_t ptrStyleEndShiftToImageDataPixel = ((regionHeight + yPos) * regionWidth + xPos) * blockByteSize + region->bufferOffset;
+						const uint8_t* texelPtr = reinterpret_cast<const uint8_t*>(data) + region->bufferOffset + (yPos * regionWidth + xPos) * asset::getTexelOrBlockBytesize(image->getCreationParameters().format);
```
better call `getTexelOrBlockBytesize` once and save to some var. This call is quite big switch so maybe better no to call it in a loop

AnastaZIuk replied:
you're right, I saved it

AnastaZIuk committed:
Stop getting texelBlockSize from function in the equation

devshgraphicsprogramming commented on src/irr/asset/IImageLoader.cpp:
```diff
@@ -0,0 +1,9 @@
+using namespace irr;
+using namespace asset;
+
+IImageLoader::~IImageLoader()
```
no cpp file needed, you can inline the constructor

devshgraphicsprogramming commented on src/irr/asset/IImageWriter.cpp:
```diff
@@ -0,0 +1,9 @@
+using namespace irr;
+using namespace asset;
+
+IImageWriter::~IImageWriter()
```
no cpp file needed, you can inline the constructor

devshgraphicsprogramming commented on src/irr/asset/IImageAssetHandlerBase.cpp:
```diff
@@ -0,0 +1,9 @@
+using namespace irr;
+using namespace asset;
+
+IImageAssetHandlerBase::~IImageAssetHandlerBase()
```
no cpp file needed, you can inline the constructor

AnastaZIuk replied:
it is a destructor as rest

devshgraphicsprogramming replied:
just put `{}` witout the `;` after the function declaration in the header (in other words, make the definition inline) you wont need a source file

AnastaZIuk replied:
I know  I know, actually it's default
